### PR TITLE
Implement LoadBalancer support in the operator

### LINKED ISSUE #1387: Allow subctl/operator --loadbalancer mode
Related to: https://github.com/submariner-io/submariner/issues/1071

Add a setting to the operator and subctl to enable load-balancer mode. 
Once enabled creates the LoadBalancer automatically in the submariner-operator namespace.

Investigate all annotations necessary on clouds to use UDP loadbalancers,  an example for AWS:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: gw1
  namespace: submariner-operator
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
spec:
  externalTrafficPolicy: Local
  selector:
    app: submariner-gateway
    gateway: active
  ports:
    - name: ipsec-encap
      protocol: UDP
      port: 4500
      targetPort: 4500
    - name: natt-discovery
      protocol: UDP
      port: 4490
      targetPort: 4490
  type: LoadBalancer

```

### PR DESCRIPTION
**Pull request by mangelajo**
Adds flag to automatically create a LoadBalancer service that points
to the active gateway.

Resolves-Issue: #1387

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

### CODE REVIEW

File: controllers/submariner/gateway_resources.go [hunk 1]
```diff
@@ -199,6 +202,11 @@ func newGatewayPodTemplate(cr *v1alpha1.Submariner) corev1.PodTemplateSpec {
 	podTemplate.Spec.Containers[0].Env = append(podTemplate.Spec.Containers[0].Env,
 		corev1.EnvVar{Name: "CE_IPSEC_FORCEENCAPS", Value: strconv.FormatBool(cr.Spec.CeIPSecForceUDPEncaps)})
 
+	if cr.Spec.LoadBalancerEnabled {
+		podTemplate.Spec.Containers[0].Env = append(podTemplate.Spec.Containers[0].Env,
+			corev1.EnvVar{Name: "PUBLIC_IP", Value: "lb:" + loadBalancerName})
```
Comment (mangelajo) [inline controllers/submariner/gateway_resources.go]:
```suggestion
			corev1.EnvVar{Name: "SUBMARINER_PUBLICIP", Value: "lb:" + loadBalancerName})
```

So we can add it to the spec struct in submariner-gateway

**Comment by mangelajo** [comment]
Ok, the upgrade job fails because of: 
```
E0607 08:37:12.738455       1 garbagecollector.go:350] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"submariner.io/v1alpha1", Kind:"Submariner", Name:"submariner", UID:"023a39ae-f49f-44e9-85e3-8f383adaa375", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"submariner-operator"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{}, deletingDependents:true, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:true, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference(nil)}: Submariner.submariner.io "submariner" is invalid: spec.loadBalancerEnabled: Required value

```

fixed now

File: controllers/submariner/gateway_resources.go [hunk 2]
```diff
@@ -70,10 +70,13 @@ func newGatewayDaemonSet(cr *v1alpha1.Submariner) *appsv1.DaemonSet {
 	return deployment
 }
 
+const appLabel = "app"
```
Comment (Jaanki) [inline controllers/submariner/gateway_resources.go:73]: I would just make this 1 const `appGatewayLabel = "\"app\": \"submariner-gateway\""`

Comment (Jaanki) [reply controllers/submariner/gateway_resources.go:73]: ohk...But the current format looks better in the next file.

File: pkg/subctl/cmd/join.go [hunk 3]
```diff
@@ -69,6 +69,7 @@ var (
 	submarinerDebug               bool
 	operatorDebug                 bool
 	labelGateway                  bool
+	loadBalancer                  bool
```
Comment (Jaanki) [inline pkg/subctl/cmd/join.go]:
```suggestion
	loadBalancerEnabled           bool
```
so the flag becomes --loadBalancerEnabled = true or false

**Review by Jaanki — CHANGES_REQUESTED** [review]
Just a suggestion. Otherwise, looks good

**Review by tpantelis — APPROVED** [review]

**Review by maayanf24 — APPROVED** [review]

File: pkg/subctl/cmd/join.go (hunk 3 above)
Comment (mangelajo) [reply pkg/subctl/cmd/join.go]: I applied your suggestion. Thanks :)

**Comment by mangelajo** [comment]
Thank you for the reviews.

File: controllers/submariner/gateway_resources.go (hunk 2 above)
Comment (mangelajo) [reply controllers/submariner/gateway_resources.go:73]: We have to use them in multiple places that's why I separated them as individual constants.

**Review by Jaanki — APPROVED** [review]
Thanks @mangelajo

**Pull request merged**